Ndev 3739

romanova-natasha opened the pull request.

Commit:
add drain wsol after test

add drain wsol after test

Commit:
add drain wsol step to fixtures

add drain wsol step to fixtures

Commit:
Merge remote-tracking branch 'origin/develop' into NDEV-3739

Commit:
add: return tokens for withdraw contract owner, drain_wsol changes: create ata if it does not exist

add: return tokens for withdraw contract owner, drain_wsol changes: create ata if it does not exist

Commit:
add use_bank=false to terraform env configuration

add use_bank=false to terraform env configuration to prevent keyError exception

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -314,12 +335,47 @@ def withdraw_contract(web3_client, faucet, accounts) -> Contract:
+) -> Contract:
     contract, _ = web3_client_sol.deploy_and_get_contract(
-        "precompiled/NeonToken", "0.8.10", account=account_with_all_tokens
+        "precompiled/NeonToken", "0.8.10", account=account_sol_chain_for_withdraw_contract
```
didn't get why do you need separated user only for deploying of the current contract , why you can't use one of existing accounts

romanova-natasha replied:
Withdraw contract requires an account with wsol to be deployed. 
I can not use `class_account_sol_chain` or `account_with_all_tokens` because I call the withdraw contract on a finalizing step in these fixtures. To avoid a recursive dependencies issue I created separate account with wsol `account_sol_chain_for_withdraw_contract`, when I finalize the withdraw contract I return wsol from `account_sol_chain_for_withdraw_contract` to bank account ata.

kristinaNikolaevaa replied:
to prevent using account_sol_chain_for_withdraw_contract in some tests or another fixtures (it's forbidden to do as you drain wsols here in current fixture) it's maybe good to move creating of the user to the withdraw_contract as well and remove account_sol_chain_for_withdraw_contract

romanova-natasha replied:
done

kristinaNikolaevaa commented on utils/evm_loader.py:
```diff
@@ -685,6 +687,46 @@ def deposit_wrapped_sol_from_solana_to_neon(self, solana_account, neon_account,
 
         self.send_token_from_solana_to_neon(solana_account, mint_pubkey, neon_account, full_amount, self.sol_chain_id)
 
+    def drain_wsol(
```
I suggest to move it from evm_loader class as you use web3_client here and it strange to pass one client to another

it is better to spilt parts connected to web3 and solana or move the method to same helpers

romanova-natasha replied:
I separated this method in two parts:
1) withdraw wsol to bank_account's ata from every account with wsol (on a finalizing account's fixtures)
2) collect tokens from bank_account's ata on a finalizing bank_account fixture

kristinaNikolaevaa commented on utils/evm_loader.py:
```diff
@@ -685,6 +687,46 @@ def deposit_wrapped_sol_from_solana_to_neon(self, solana_account, neon_account,
+        assert ata_balance_after == ata_balance_before + value
+
+        balance_return_account = self.get_solana_balance(return_to.pubkey())
+        print("Balance bank account before return: ", balance_return_account)
```
please remove all prints (or use allure steps instead)

romanova-natasha replied:
done

kristinaNikolaevaa commented on utils/evm_loader.py:
```diff
@@ -685,6 +687,46 @@ def deposit_wrapped_sol_from_solana_to_neon(self, solana_account, neon_account,
+        spl_token = SplToken(self, WRAPPED_SOL_MINT, TOKEN_PROGRAM_ID, withdraw_to)
+        ata_balance_before = int(spl_token.get_balance(ata, commitment=Confirmed).value.amount)
+        ata_info = spl_token.get_account_info(ata, commitment=Confirmed)
+        assert ata_info.owner == withdraw_to.pubkey()
```
why do you need this check? you check here the work of solana and method  
self.create_associate_token_acc(withdraw_to, withdraw_to, WRAPPED_SOL_MINT)

romanova-natasha replied:
removed

kristinaNikolaevaa commented on utils/evm_loader.py:
```diff
@@ -685,6 +687,46 @@ def deposit_wrapped_sol_from_solana_to_neon(self, solana_account, neon_account,
+
+        tx = web3_client.make_raw_tx(from_=withdraw_from, amount=neon_balance)
+        value = (neon_balance - web3_client.eth.estimate_gas(tx) * web3_client.gas_price()) // 10**9
+        tx["value"] = value * 10**9
```
did't get the logic 
first you do value // 10**9
and next line is  value * 10**9

romanova-natasha replied:
withdraw contract requires value to be divisible to 10**9
I drop remaining amount with // operation
Idk maybe it could be done better

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -314,12 +335,47 @@ def withdraw_contract(web3_client, faucet, accounts) -> Contract:
+    if environment.use_bank:
+        evm_loader.drain_wsol(
+            withdraw_from=account_sol_chain_for_withdraw_contract,
+            withdraw_to=solana_account,
```
why do you need to use solana_account ? now you do account_sol_chain_for_withdraw_contract -> solana_account -> bank_account

But you can simplify it as  account_sol_chain_for_withdraw_contract  -> bank_account

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/conftest.py:
Can you please add draining wsol for neon_user as well?
I see we use deposit_wrapped_sol_from_solana_to_neon() for it in several tests

romanova-natasha replied:
neon_user does not have an eth signature, do you want to do a new return method with scheduled tx?

Commit:
fix finalizing of the fixture withdraw_contract_sol_chain

fix finalizing of the fixture withdraw_contract_sol_chain

Commit:
refactoring according to the review

refactoring according to the review

Commit:
fix typo

fix typo

Commit:
add withdraw wsol from neon_user via scheduled trx

add withdraw wsol from neon_user via scheduled trx

Commit:
add test to reproduce an issue with scheduled execution of withdraw contract function

add test to reproduce an issue with scheduled execution of withdraw contract function

Commit:
fix insufficient token for transfer error

fix insufficient token for transfer error

Commit:
fix evm tests: add a separate fixture of the neon_user

fix evm tests: add a separate fixture of the neon_user

Commit:
skip tests due to NDEV-3795

skip tests due to NDEV-3795

romanova-natasha committed:
Merge branch 'develop' into NDEV-3739

kristinaNikolaevaa commented on conftest.py:
```diff
@@ -332,8 +349,27 @@ def neon_user_no_sols(pytestconfig, bank_account, faucet, environment) -> NeonUs
 
 
+@pytest.fixture(scope="function")
+def neon_user_for_evm_test(
```
You can leave single neon_user because for evm tests run there are no bank account (and it's imposible run on devnet) and evm_loader.drain_sol will not be run

romanova-natasha replied:
it is not about `bank_account`
it is about `withdraw_contract_sol_chain` fixture
this contract is deployed with proxy and it affects evm test run

kristinaNikolaevaa replied:
then you can owerrride neon_user in evm test conftest and move this one to basic tests conftest

romanova-natasha replied:
done

kristinaNikolaevaa commented on conftest.py:
```diff
@@ -344,7 +380,20 @@ def bank_account(pytestconfig: Config) -> Generator[Keypair | None, None, None]:
         account = Keypair.from_bytes(key)
+
+    if pytestconfig.environment.use_bank:
+        ata = sol_client_session.create_associate_token_acc(account, account, WRAPPED_SOL_MINT)
```
you already have create_associate_token_acc() call inside withdraw_neon_to_solana_sol_sign

romanova-natasha replied:
`create_associate_token_acc` - returns ata (if it exists) or creates it if it is not
maybe we should rename it not to be confused all the time?

kristinaNikolaevaa replied:
maybe 
to get ata there is a separate function get_associated_token_address(owner.pubkey(), token_mint)

romanova-natasha replied:
I use get_associated_token_address in withdraw_neon_to_solana_sol_sign method now. I create an ata for bank_account here

kristinaNikolaevaa commented on integration/tests/basic/solana_signature/test_send_scheduled_transactions.py:
```diff
@@ -1,16 +1,37 @@
+            int(1 * LAMPORT_PER_SOL),
+        )
+        recipient = Keypair()
+        evm_loader.request_airdrop(recipient.pubkey(), 5 * LAMPORT_PER_SOL, commitment=Confirmed)
```
it will not pass on devnet (use perepared accounts)

romanova-natasha replied:
done

kristinaNikolaevaa commented on integration/tests/basic/solana_signature/test_send_scheduled_transactions.py:
```diff
@@ -1,16 +1,37 @@
+        )
+
+        balance_after = web3_client_sol.get_balance(neon_user.checksum_address) // 10**9
+        assert balance_after == 0
```
it's imposible it will be ==0 
estimation value != real value of the transaction , more likely it will be less and part of tokens will be returned after transaction execution

romanova-natasha replied:
I left a check: balance_after != balance_before

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -314,12 +323,32 @@ def withdraw_contract(web3_client, faucet, accounts) -> Contract:
+) -> Contract:
+    account = web3_client_session.create_account_with_balance(faucet, bank_account=eth_bank_account)
+    if environment.use_bank:
+        evm_loader.send_sol(bank_account, solana_account.pubkey(), int(1 * LAMPORT_PER_SOL))
```
doesn't solana_account  have enough sols?

romanova-natasha replied:
extra 1 sol is added to solana_account, fixed

kristinaNikolaevaa commented on integration/tests/conftest.py:
```diff
@@ -314,12 +323,32 @@ def withdraw_contract(web3_client, faucet, accounts) -> Contract:
+    else:
+        evm_loader.request_airdrop(solana_account.pubkey(), 1 * LAMPORT_PER_SOL)
+
+    evm_loader.deposit_wrapped_sol_from_solana_to_neon(
```
why do you need it here?

romanova-natasha replied:
I need tokens to deploy contract with `account`. 
It is a separate account to deploy withdraw contract, cause we use this contract for every account with wsol to refund it in the and of every run. To avoid recursive dependency I created a separate one inside this fixture.

kristinaNikolaevaa commented on utils/helpers.py:
```diff
@@ -324,3 +328,49 @@ def decode_error_output(data_hex):
+        withdraw contract requires trx value to be divisible to 10**9
+        remainings of the value are dropped with // operation
+    """
+    value = (amount - web3_client.eth.estimate_gas(tx) * web3_client.gas_price()) // 10**9
```
web3_client.eth.estimate_gas(tx) here returns  value for empty trx(as tx has no data field) , not for real one with withdraw_on_chain() call
Not sure it will work on devnet

romanova-natasha replied:
I'll try to check on our devnet (if it works fine)

romanova-natasha replied:
I did check it on our devnet. Withdraw was successful, but I am also not sure about stability of the code so I added a multiplier=1.1 for gas.

kristinaNikolaevaa replied:
it is better ro estimate correct transaction ( with data and 'to' field)
to have here random value as estimated gas is dangerous

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on utils/helpers.py:
```diff
@@ -324,3 +328,49 @@ def decode_error_output(data_hex):
+    assert ata_balance_after >= ata_balance_before + trx_estimate_obj.value // 10**9
+
+    balance_withdraw_from_after = web3_client_sol.get_balance(withdraw_from.checksum_address)
+    assert balance_withdraw_from_after == 0
```
the same, unlikly it will be==0

romanova-natasha replied:
just left the check: balance_withdraw_from_after != balance_withdraw_from_before

kristinaNikolaevaa commented on utils/helpers.py:
```diff
@@ -324,3 +328,49 @@ def decode_error_output(data_hex):
+    """
+    trx_estimate_obj.value = ((trx_estimate_obj.value - gas) // 10**9) * 10**9
+
+    if trx_estimate_obj.value > 0:
```
strange condition
you can move it up, check amount and no to do estimation for the case as well

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on conftest.py:
```diff
@@ -344,7 +380,20 @@ def bank_account(pytestconfig: Config) -> Generator[Keypair | None, None, None]:
+        spl_token = SplToken(sol_client_session, WRAPPED_SOL_MINT, TOKEN_PROGRAM_ID, account)
+        ata_balance = int(spl_token.get_balance(ata, commitment=Confirmed).value.amount)
+        spl_token.close_account(account=ata, dest=account.pubkey(), authority=account)
+        wait_condition(
```
I don't think wee need checks like that here (as close_account is solana functionality)
The check can fail as we use the same bank account for all threads in tests run and balance is changing in parallel trx

romanova-natasha replied:
removed

Commit:
fix test for devnet run

fix test for devnet run

Commit:
fix review

fix review

Commit:
Merge branch 'NDEV-3739' of https://github.com/neonlabsorg/neon-tests into NDEV-3739

Commit:
add gas multiplier to withdraw trx value

add gas multiplier to withdraw trx value

romanova-natasha committed:
Merge branch 'develop' into NDEV-3739

Commit:
fix review

fix review

Commit:
Merge remote-tracking branch 'origin/develop' into NDEV-3739
